src/type-data-frame.c

```diff
@@ -39,6 +39,84 @@ SEXP new_data_frame(SEXP x, R_len_t n) {
   return x;
 }
 
+static R_len_t pull_data_frame_size(SEXP x, SEXP n);
+static void set_data_frame_class(SEXP x, SEXP cls);
+
+// [[ register() ]]
+SEXP vctrs_new_data_frame(SEXP x, SEXP n, SEXP attributes, SEXP cls) {
+  if (TYPEOF(x) != VECSXP) {
+    Rf_errorcall(R_NilValue, "`x` must be a list");
+  }
+
+  R_len_t size = pull_data_frame_size(x, n);
```

Comment:
What does `pull_` mean? Why is this not consistent with `df_attr()` naming? To avoid these questions, maybe just inline the function since it's used only once?

Reply:
Inlining makes it _slightly_ harder to read, I think. So I've just renamed it.

Inlined is probably:

```c
  R_len_t size;
  if (n == R_NilValue) {
    size = df_raw_size_from_list(x);
  } else {
    if (TYPEOF(n) != INTSXP || Rf_length(n) != 1) {
      Rf_errorcall(R_NilValue, "`n` must be an integer of size 1");
    }
    size = r_int_get(n, 0);
  }
```